internal/compiler/passes/remove_return.rs

```diff
@@ -0,0 +1,378 @@
+                if iter.peek().is_none() {
+                    return ExpressionResult::ReturnObject {
+                        value: codeblock_with_expr(stmts, value),
+                        has_value,
+                        has_return_value,
+                    };
+                };
+                return continue_codeblock(
+                    iter,
+                    ty,
+                    ctx,
+                    value,
+                    stmts,
+                    has_return_value,
+                    has_value,
+                );
+            }
+        }
+    }
+    ExpressionResult::Just(Expression::CodeBlock(stmts))
+}
+
+fn continue_codeblock(
+    iter: std::iter::Peekable<impl Iterator<Item = Expression>>,
+    ty: &Type,
+    ctx: &RemoveReturnContext,
+    return_object: Expression,
+    mut stmts: Vec<Expression>,
+    has_return_value: bool,
+    has_value: bool,
+) -> ExpressionResult {
+    let rest = process_codeblock(iter, ty, ctx).into_return_object(ty, &ctx.ret_ty);
+    static COUNT: std::sync::atomic::AtomicUsize = std::sync::atomic::AtomicUsize::new(0);
+    let unique_name =
+        format!("return_check_merge{}", COUNT.fetch_add(1, std::sync::atomic::Ordering::Relaxed));
+    let load = Box::new(Expression::ReadLocalVariable {
+        name: unique_name.clone(),
+        ty: return_object.ty(),
+    });
+    stmts.push(Expression::StoreLocalVariable { name: unique_name, value: return_object.into() });
+    stmts.push(Expression::Condition {
+        condition: Expression::StructFieldAccess {
+            base: load.clone(),
+            name: FIELD_CONDITION.into(),
+        }
+        .into(),
+        true_expr: rest.into(),
+        false_expr: ExpressionResult::Return(has_return_value.then(|| {
+            Expression::StructFieldAccess { base: load.clone(), name: FIELD_RETURNED.into() }
+        }))
+        .into_return_object(ty, &ctx.ret_ty)
+        .into(),
+    });
+    ExpressionResult::ReturnObject {
+        value: Expression::CodeBlock(stmts),
+        has_value,
+        has_return_value,
+    }
+}
+
+struct RemoveReturnContext {
+    ret_ty: Type,
+}
+
+#[derive(Debug)]
+enum ExpressionResult {
+    /// The expression maps directly to a LLR expression
+    Just(Expression),
+    /// The expression used `return` so we need to check for the return slot
+    MaybeReturn {
+        /// Some statements that initializes some temporary variable (eg arguments to something called later)
+        pre_statements: Vec<Expression>,
+        /// Boolean expression: false means return
+        condition: Expression,
+        /// Value being returned if condition is false
+        returned_value: Option<Expression>,
+        /// The value when we don't return
+        actual_value: Option<Expression>,
+    },
+    /// The expression returns unconditionally
+    Return(Option<Expression>),
+    /// The expression is of type `{ condition: bool, actual: ty, returned: ret_ty}`
```

Comment:
Maybe it's self-explaining, but I had to double check that the condition implies the use the actual field and not that it means "has_return_value".

```suggestion
    /// The expression is of type `{ condition_to_use_actual: bool, actual: ty, returned: ret_ty}`
```